spec/fastsync/version2/fsv2_scheduler.tla

```diff
@@ -0,0 +1,532 @@
+  /\ UNCHANGED envRunning
+
+OnPcBlockVerificationFailure ==
+  /\ inEvent' \in [type: {"pcBlockVerificationFailure"}, firstPeerID: PeerIDs, secondPeerID: PeerIDs]
+  /\ UNCHANGED envRunning
+
+\* messages from scheduler
+OnScFinishedEv ==
+  /\ inEvent' = [type |-> "tNoAdvanceExp"]
+  /\ envRunning' = FALSE \* stop the env
+
+NextEnv ==
+  IF ~envRunning
+    THEN inEvent' = noEvent /\ UNCHANGED envRunning
+  ELSE
+    \/ OnScFinishedEv
+    \/ OnGlobalTimeoutTicker
+    \/ OnAddPeerEv
+    \/ OnTrySchedule
+    \/ OnStatusResponseEv
+    \/ OnBlockResponseEv
+    \/ OnRemovePeerEv
+    \/ OnPcBlockProcessed
+    \/ OnPcBlockVerificationFailure
+
+(* ----------------------------------------------------------------------------------------------*)
+(* The system is the composition of the environment and the schedule                             *)
+(* ----------------------------------------------------------------------------------------------*)
+Init == turn = "environment" /\ InitEnv /\ InitSc
+
+FlipTurn == 
+  turn' = (
+    IF turn = "scheduler" THEN 
+      "environment" 
+    ELSE
+      "scheduler"
+  ) 
+
+Next == \* scheduler and environment alternate their steps (synchronous composition introduces more states)
+  /\ FlipTurn
+  /\ IF turn = "scheduler" THEN 
+       /\ NextSc
+       /\ inEvent' = noEvent
+       /\ UNCHANGED envRunning
+     ELSE
+       /\ NextEnv
+       /\ outEvent' = noEvent
+       /\ UNCHANGED <<scRunning, scheduler>>  
+
+(* ----------------------------------------------------------------------------------------------*)
+(* Expected properties                                                                           *)
+(* ----------------------------------------------------------------------------------------------*)
+
+TypeOK ==
+    /\ turn \in {"scheduler", "environment"}
+    /\ inEvent \in InEvents
+    /\ envRunning \in BOOLEAN
+    /\ outEvent \in OutEvents
+    /\ scheduler \in [
+         initHeight: Heights \cup {ultimateHeight + 1},
+         height: Heights \cup {ultimateHeight + 1},
+         peers: SUBSET PeerIDs,
+         peerHeights: [PeerIDs -> Heights \cup {None}],
+         peerStates: [PeerIDs -> PeerStates],
+         blocks: SUBSET Heights,
+         blockStates: [Heights -> BlockStates],
+         pendingBlocks: [Heights -> PeerIDs \cup {None}],
+         receivedBlocks: [Heights -> PeerIDs \cup {None}]
+       ]
+       
+TerminationWhenNoAdvance ==
+  (WF_turn(FlipTurn) /\ <> (inEvent.type = "tNoAdvanceExp"))
+    => <> ~scRunning
+
+NoFailuresAndTimeouts ==
+    \* no peer removal
+    /\ inEvent.type /= "peerRemoveEv"
+    \* no invalid blocks
+    /\ inEvent.type /= "bcNoBlockResponse"
+    /\ inEvent.type /= "pcBlockVerificationFailure"
+NoGlobalTimeout == inEvent.type /= "tNoAdvanceExp"
+
+FiniteResponse ==
+    <>[] (inEvent.type
+         \in {"bcAddNewPeer", "bcStatusResponse", "bcBlockResponse", "pcBlockProcessed", "pcBlockVerificationFailure", "rTrySchedule"})
```

Comment:
Can we clarify what this is, ie. how/why certain inEvents are left out?

Reply:
Will add comments.